java/arcs/core/storage/WriteBack.kt

```diff
@@ -0,0 +1,145 @@
+import kotlinx.coroutines.channels.Channel
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+
+/**
+ * A layer to decouple local data updates and underlying storage layers flushes.
+ *
+ * The modern storage stacks including VFS page dirty write-back, io-scheduler
+ * merging and re-ordering, flash device turbo-write and/or buffer flush, etc
+ * are all implemented in an efficient way where no data/metadata is written
+ * through to storage media for every single write operation unless explicitly
+ * being requested.
+ */
+interface WriteBack {
+    /**
+     * Write-through: flush directly all data updates to the next storage layer.
+     */
+    suspend fun flush(job: suspend () -> Unit)
+    /**
+     * Write-back: queue up data updates and let write-back threads decide how and
+     * when to flush all data updates to the next storage layer.
+     */
+    suspend fun asyncFlush(job: suspend () -> Unit)
+
+    /** Await completion of all active flush jobs. */
+    suspend fun awaitIdle()
+}
+
+/** The factory interfaces of [WriteBack] implementations. */
+interface WriteBackFactory {
+    fun create(
+        protocol: String = "",
+        /**
+         * Provide a dedicated write-back thread pool or leave it to implementations
+         * to decide what threads to be designated as the write-backers.
+         */
+        writebackThreads: ExecutorService? = null,
+        /**
+         * The maximum queue size above which new incoming flush jobs will be suspended.
+         */
+        queueSize: Int = 10
+    ): WriteBack
+}
+
+/** Write-back implementation for Arcs Stores.*/
+class StoreWriteBack private constructor(
+    protocol: String,
+    writebackThreads: ExecutorService?,
+    queueSize: Int
+) : WriteBack,
+    Channel<suspend () -> Unit> by Channel(queueSize),
+    CoroutineScope by CoroutineScope(
+        writebackThreads?.asCoroutineDispatcher() ?: Dispatchers.IO
+    ),
+    Mutex by Mutex() {
+    // Only apply write-back to physical storage media(s).
+    private val passThrough = protocol != Protocols.DATABASE_DRIVER
+
+    // The number of active flush jobs.
+    private var activeJobs = 0
+
+    // The signal to block/release who are waiting for completion of active flush jobs.
+    private val awaitSignal = Mutex()
+
+    init {
+        // One of write-back thread(s) will wake up and execute flush jobs in FIFO order
+        // when there are pending flush jobs in queue. Powerful features like batching,
+        // merging, filtering, etc can be implemented at this call-site in the future.
+        if (!passThrough) {
+            launch {
+                try {
+                    while (true) {
+                        exitFlushSection { receive()() }
+                    }
+                } finally {
+                    if (awaitSignal.isLocked) {
+                        awaitSignal.unlock()
+                    }
+                }
+            }
```

Comment:
This behavior may be easier to accomplish safely using `Flow`s, rather than low-level Channel semantics.

Reply:
Flow looks more better to use with cold stream or fixed data sets.
And I'd like a StoreWriteBack as-is a Channel. :)

Reply:
I finally decided to follow your suggestion to use flow with hot stream. Thanks.